manifests/install.pp

```diff
@@ -14,86 +14,81 @@
-        ensure  => file,
-        mode    => '0755',
-        content => template('puppet_agent/do_install.sh.erb')
-      }
-      -> exec { 'solaris_install script':
-        command => "/usr/bin/ctrun -l none ${_installsh} ${::puppet_agent_pid} 2>&1 > ${_logfile} &",
-      }
+  # Solaris, MacOS, Windows platforms will require more than just a package
+  # resource to install correctly. These will call to other classes that
+  # define how the installations work.
+  if $::operatingsystem == 'Solaris' {
+    class { 'puppet_agent::install::solaris':
+      package_version => $package_version,
+      install_options => $install_options,
     }
+    contain '::puppet_agent::install::solaris'
   } elsif $::operatingsystem == 'Darwin' {
+    # Prevent re-running the script install
     if $puppet_agent::aio_upgrade_required {
-      $install_script = 'osx_install.sh.erb'
-
-      $_logfile = "${::env_temp_variable}/osx_install.log"
-      notice("Puppet install log file at ${_logfile}")
-
-      $_installsh = "${::env_temp_variable}/osx_install.sh"
-      file { "${_installsh}":
-        ensure  => file,
-        mode    => '0755',
-        content => template('puppet_agent/do_install.sh.erb')
-      }
-      -> exec { 'osx_install script':
-        command => "${_installsh} ${::puppet_agent_pid} 2>&1 > ${_logfile} &",
+      class { 'puppet_agent::install::darwin':
+        package_version => $package_version,
+        install_options => $install_options,
       }
+      contain '::puppet_agent::install::darwin'
     }
   } elsif $::osfamily == 'windows' {
     # Prevent re-running the batch install
     if $puppet_agent::aio_upgrade_required {
-      class { 'puppet_agent::windows::install':
+      class { 'puppet_agent::install::windows':
         install_dir     => $install_dir,
         install_options => $install_options,
       }
-    }
-  } elsif ($::osfamily == 'Debian') and ($package_version != 'present') {
-    package { $::puppet_agent::package_name:
-      ensure          => "${package_version}-1${::lsbdistcodename}",
-      install_options => $install_options,
+      contain '::puppet_agent::install::windows'
     }
   } else {
+    if $::operatingsystem == 'AIX' {
+      # AIX installations always use RPM directly since no there isn't any default package manager for rpms
+      $_package_version = $package_version
+      $_install_options = concat(['--ignoreos'],$install_options)
+      $_provider = 'rpm'
+      $_source = "${::puppet_agent::params::local_packages_dir}/${::puppet_agent::prepare::package::package_file_name}"
+    } elsif $::osfamily == 'Debian' {
+      $_install_options = $install_options
+      if $::puppet_agent::absolute_source {
+        # absolute_source means we use dpkg on debian based platforms
+        $_package_version = 'present'
+        $_provider = 'dpkg'
+        $_source = "${::puppet_agent::params::local_packages_dir}/${::puppet_agent::prepare::package::package_file_name}"
```

Comment:
Should be `$_source = $::puppet_agent::absolute_source`?

Not sure package would get to `"${::puppet_agent::params::local_packages_dir}/${::puppet_agent::prepare::package::package_file_name}"` unless the `puppet_agent::prepare::package` class was also used for Debian/RHEL/SUSE/etc.

Reply:
Ah, actually the second comment is what's missing. In cases of absolute source debian/rhel/suse should be using prepare::package.